File: 백준/Gold/31427. 신촌 도로망 관리와 쿼리/신촌 도로망 관리와 쿼리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.StringTokenizer;

public class Main {

    static final int SIZE = 5;

    static int N, M, Q;
    static List<List<Node>> edges = new ArrayList<>();
    static int[] parent;
    static int[] perm = new int[SIZE];
    static Map<Temp, int[]> calc = new HashMap<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        Q = Integer.parseInt(st.nextToken());
        parent = new int[N + 1];
        perm = new int[SIZE];
        for (int i = 0; i <= SIZE; i++) {
            edges.add(new ArrayList<>());
        }
        for (int i = 0; i < SIZE; i++) {
            perm[i] = i;
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int c = st.nextToken().charAt(0) - 'A';
            edges.get(c).add(new Node(u, v, c));
        }

        do {
            for (int i = 1; i <= N; i++) {
                parent[i] = i;
            }
            int[] c = new int[SIZE];
            for (int i = 0; i < SIZE; i++) {
                for (Node edge : edges.get(perm[i])) {
                    if (findParent(edge.u) != findParent(edge.v)) {
                        union(edge.u, edge.v);
                        c[edge.w]++;
                    }
                }
            }
            calc.put(new Temp(perm), c);
        } while (np());

        List<Integer> order = new ArrayList<>();
        for (int i = 0; i < SIZE; i++) {
            order.add(i);
        }
        int[] val = new int[SIZE];
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < Q; i++) {
            long ans = 0;
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < SIZE; j++) {
                val[j] = Integer.parseInt(st.nextToken());
            }

            Collections.sort(order, (a, b) -> Integer.compare(val[a], val[b]));

            int[] result = calc.get(new Temp(order));
            for (int j = 0; j < SIZE; j++) {
                ans += (long) val[order.get(j)] * result[order.get(j)];
            }
            sb.append(ans).append("\n");
        }
        System.out.println(sb);
    }

    private static int findParent(int a) {
        if (a == parent[a]) {
            return a;
        }
        return parent[a] = findParent(parent[a]);
    }

    private static void union(int a, int b) {
        int p1 = findParent(a);
        int p2 = findParent(b);

        if (p1 != p2) {
            parent[p2] = p1;
        }
    }

    private static boolean np() {
        int i = SIZE - 1;

        while (i > 0 && perm[i - 1] >= perm[i]) {
            i--;
        }

        if (i == 0) {
            return false;
        }

        int j = SIZE - 1;
        while (perm[i - 1] >= perm[j]) {
            j--;
        }
        swap(i - 1, j);

        int k = SIZE - 1;
        while (i < k) {
            swap(i++, k--);
        }
        return true;
    }

    private static void swap(int i, int j) {
        perm[i] ^= perm[j];
        perm[j] ^= perm[i];
        perm[i] ^= perm[j];
    }

    static class Temp {

        int[] order = new int[SIZE];

        public Temp(int[] order) {
            for (int i = 0; i < SIZE; i++) {
                this.order[i] = order[i];
            }
        }

        public Temp(List<Integer> list) {
            for (int i = 0; i < SIZE; i++) {
                this.order[i] = list.get(i);
            }
        }

        @Override
        public boolean equals(Object object) {
            if (this == object) {
                return true;
            }
            if (object == null || getClass() != object.getClass()) {
                return false;
            }
            Temp temp = (Temp) object;
            return Arrays.equals(order, temp.order);
        }

        @Override
        public int hashCode() {
            return Arrays.hashCode(order);
        }

        @Override
        public String toString() {
            return "Temp{" +
                    "order=" + Arrays.toString(order) +
                    '}';
        }
    }

    static class Node {

        int u;
        int v;
        int w;

        public Node(int u, int v, int w) {
            this.u = u;
            this.v = v;
            this.w = w;
        }

        @Override
        public String toString() {
            return "Node{" +
                    "u=" + u +
                    ", v=" + v +
                    ", w=" + w +
                    '}';
        }
    }
}

File: 백준/Gold/31427. 신촌 도로망 관리와 쿼리/신촌 도로망 관리와 쿼리.cc
#include<bits/stdc++.h>
using namespace std;

#define SIZE 5

class Node {
public:
    int u, v, w;
    Node() {} // 기본 생성자 추가
    Node(int u, int v, int w) : u(u), v(v), w(w) {}
};

class Temp {
public:
    vector<int> order;
    Temp(vector<int>& order) : order(order) {}
    bool operator==(const Temp& other) const { return order == other.order; }
};

namespace std {
    template<>
    struct hash<Temp> {
        size_t operator()(const Temp& t) const {
            return hash<int>()(accumulate(t.order.begin(), t.order.end(), 0));
        }
    };
}

vector<vector<Node>> edges(SIZE);
unordered_map<Temp, vector<Node>> calc;
vector<int> parent, perm(SIZE);
int N, M, Q;

int findParent(int a) {
    if (a == parent[a]) return a;
    return parent[a] = findParent(parent[a]);
}

void unionParent(int a, int b) {
    int p1 = findParent(a);
    int p2 = findParent(b);
    if (p1 != p2) parent[p2] = p1;
}

bool next_permutation() {
    int i = SIZE - 1;
    while (i > 0 && perm[i - 1] >= perm[i]) i--;
    if (i == 0) return false;
    int j = SIZE - 1;
    while (perm[i - 1] >= perm[j]) j--;
    swap(perm[i - 1], perm[j]);
    int k = SIZE - 1;
    while (i < k) swap(perm[i++], perm[k--]);
    return true;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> N >> M >> Q;
    parent.resize(N + 1);
    for (int i = 0; i < SIZE; i++) perm[i] = i;
    for (int i = 0; i < M; i++) {
        int u, v;
        char c;
        cin >> u >> v >> c;
        edges[c - 'A'].push_back(Node(u, v, c - 'A'));
    }

    do {
        for (int i = 1; i <= N; i++) parent[i] = i;
        vector<Node> c(N - 1);
        int idx = 0;
        for (int i = 0; i < SIZE; i++) {
            for (Node& edge : edges[perm[i]]) {
                if (findParent(edge.u) != findParent(edge.v)) {
                    unionParent(edge.u, edge.v);
                    c[idx++] = edge;
                }
            }
        }
        calc[Temp(perm)] = c;
    } while (next_permutation());

    vector<int> order(SIZE), val(SIZE);
    for (int i = 0; i < SIZE; i++) order[i] = i;
    for (int i = 0; i < Q; i++) {
        long long ans = 0;
        for (int j = 0; j < SIZE; j++) cin >> val[j];
        sort(order.begin(), order.end(), [&](int a, int b) { return val[a] < val[b]; });

        vector<Node>& nodes = calc[Temp(order)];
        for (Node& node : nodes) ans += val[node.w];
        cout << ans << "\n";
    }

    return 0;
}
